python_modules/dagster/dagster/_core/errors.py

```diff
@@ -68,6 +68,52 @@ class DagsterInvalidDeserializationVersionError(DagsterError):
     """Indicates that a serialized value has an unsupported version and cannot be deserialized."""
 
 
+PYTHONIC_CONFIG_ERROR_VERBIAGE = """
+This value can be a:
+    - Python primitive type
+        - int, float, bool, str, list
+    - A Python Dict or List type containing other valid types
+    - Custom data classes extending dagster.Config
+    - A Pydantic discriminated union type
+"""
+
+PYTHONIC_RESOURCE_ADDITIONAL_TYPES = """
+\nIf this value represents a resource dependency, its annotation must either:
+    - Extend dagster.ConfigurableResource, dagster.ConfigurableIOManager, or
+    - Be wrapped in a ResourceDependency annotation, e.g. ResourceDependency[GitHub]
+"""
+
+
+class DagsterInvalidPythonicConfigDefinitionError(DagsterError):
+    """Indicates that you have attempted to construct a Pythonic config or resource class with an invalid value.
+    """
+
+    def __init__(
+        self,
+        config_class: Optional[Type],
+        field_name: Optional[str],
+        invalid_type: Any,
+        is_resource: bool = False,
+        **kwargs,
+    ):
+        self.invalid_type = invalid_type
+        self.field_name = field_name
+        self.config_class = config_class
+        super(DagsterInvalidPythonicConfigDefinitionError, self).__init__(
+            (
+                "Error defining Dagster config class{config_class}{field_name}.\nUnable"
+                " to resolve config type {invalid_type}. \n\n{PYTHONIC_CONFIG_ERROR_VERBIAGE}"
+            ).format(
+                config_class=f" {repr(config_class)}" if config_class else "",
+                field_name=f"on field '{field_name}'" if field_name else "",
+                invalid_type=repr(invalid_type),
+                PYTHONIC_CONFIG_ERROR_VERBIAGE=PYTHONIC_CONFIG_ERROR_VERBIAGE
+                + (PYTHONIC_RESOURCE_ADDITIONAL_TYPES if is_resource else ""),
+            ),
+            **kwargs,
```

Comment:
minor nit. This could be more clear in its own function with `"""` so you don't have to deal with `\n`